src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const newStates = [];
+
+  actions.forEach((action) => {
+    switch (action.type) {
+      case 'addProperties':
+        const StateBfAdd = { ...(newStates[newStates.length - 1] || state) };
+
+        newStates.push({ ...StateBfAdd, ...action.extraData });
+        break;
+      case 'removeProperties':
+        const stateBfRemove = { ...(newStates[newStates.length - 1] || state) };
+
+        action.keysToRemove.forEach((key) => {
+          if (Object.keys(stateBfRemove).includes(key)) {
+            delete stateBfRemove[key];
+          }
+        });
+        newStates.push({ ...stateBfRemove });
+        break;
+      case 'clear':
+        newStates.push({});
+        break;
+      default:
+        newStates.push({ ...state });
```

Comment:
[CODE STYLE]: Having a default case in the switch is good for error handling, but pushing the unchanged state is not an appropriate default action. Consider throwing an error or handling unexpected action types differently.